Capture shader modules

ilkkasaa opened the pull request:
Captures shader modules and prints the disassembled
SPIR-V to the output Amber file.

New intercepted Vulkan commands:
- vkCreateGraphicsPipelines
- vkCreateShaderModule
- vkDestroyShaderModule

ilkkasaa committed:
Capture shader modules

Captures shader modules and prints the disassembled
SPIR-V to the output Amber file.

New intercepted vulkan commands:
- vkCreateGraphicsPipelines
- vkCreateShaderModule
- vkDestroyShaderModule

ilkkasaa committed:
Rename some variables in vk_deep_copy.h and .cc

ilkkasaa committed:
Fix format

paulthomson committed:
Remove use of "spirv_constant.h" and more

Added gf_layers_layer_util/spirv.h.

Remove a few uses of auto.

paulthomson committed:
Fix include style

paulthomson commented on src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/amber_scoop_layer.h:
```diff
@@ -62,6 +66,9 @@ struct DeviceData {
   // Tracked device data:
 
   ProtectedMap<VkCommandBuffer, CommandBufferData> command_buffers_data;
+  ProtectedMap<VkPipeline, VkGraphicsPipelineCreateInfo> graphics_pipelines;
+  ProtectedMap<VkShaderModule, std::unique_ptr<ShaderModuleData>>
+      shader_modules_data;
```
I think we should have a `GraphicsPipelineData` class, which will take a `const VkGraphicsPipelineCreateInfo&` in the constructor, which will be deep copied, and the destructor will `DeepDelete` it.

I think `ShaderModuleData` should similarly take a `const VkShaderModuleCreateInfo&` in its constructor and `DeepCopy` it.

Basically, I think all (most?) calls to `DeepCopy` should be in a constructor, with `DeepDelete` in the corresponding destructor.

ilkkasaa replied:
That makes sense. I think there will be great amount of those classes with only create info as a field and constructor and destructor implemented (and getter for the create info), so should we add all those in a single .h file or one per class? Or maybe even use a template class like:
```c++
template <typename CreateInfoType>
class CreateInfoWrapper {
 public:
  explicit CreateInfoWrapper(const CreateInfoType &create_info) : create_info_(DeepCopy(create_info)) {}

  ~CreateInfoWrapper() {
    DeepDelete(create_info_);
  }

  CreateInfoType* GetCreateInfo() {
    return &create_info_;
  }

 private:
  CreateInfoType create_info_;
};
```

paulthomson replied:
I don't mind. Up to you.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -25,6 +25,7 @@
 
 #include "VkLayer_GF_amber_scoop/command_buffer_data.h"
 #include "VkLayer_GF_amber_scoop/draw_call_tracker.h"
+#include "VkLayer_GF_amber_scoop/vk_deep_copy.h"
```
I think we should be able to remove the vk_deep_copy.h include

paulthomson commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -177,6 +178,96 @@ VKAPI_ATTR void VKAPI_CALL vkCmdDrawIndexed(
+      device, pipelineCache, createInfoCount, pCreateInfos, pAllocator,
+      pPipelines);
+
+  if (result == VK_SUCCESS) {
```
Change to: `if (result != VK_SUCCESS) { return result; }` to reduce indentation.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -177,6 +178,96 @@ VKAPI_ATTR void VKAPI_CALL vkCmdDrawIndexed(
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+      device_data->graphics_pipelines.Put(pPipelines[i], create_info);
+
+      // Register the pipeline as a user of shader module of each stage so the
+      // shader modules can be freed from the memory when no longer used by any
+      // pipeline.
```
Some alternatives:

1. We could simply add a copy of the `ShaderModuleData` into the `GraphicsPipelineData` object (assuming you add `GraphicsPipelineData` as I suggested). But perhaps games are likely to create many pipelines with the same shaders (and keep them around), which could take up large amounts of memory.
2. You could just increment a reference count on the ShaderModuleData, instead of storing the list of pipelines. Then, when a pipeline is destroyed, you can iterate over the shader modules in the `VkGraphicsPipelineCreateInfo` and deref them. 

Actually, there is another problem with the current solution and solution 2. A game creates a shader module, uses it to create a pipeline, destroys the shader module, creates a new shader module. The new shader module might get the same handle as the destroyed handle. Thus, I suggest you try solution 1 for now.

ilkkasaa replied:
I fixed this with solution 1, but I used shared pointers to avoid copying shader module data.

paulthomson replied:
Nice!

paulthomson reviewed (commented):
Looks good! I made a few minor changes. I prefer to only use `auto` when the type is already present (e.g. on the right hand side) or has a very long name (like iterator types). I left some comments that I think are important.

ilkkasaa committed:
Add CreateInfoWrapper class

Adds a class for automatic deep copy/delete of vulkan create info
structs.

Minor fixes and changes.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/vk_deep_copy.cc:
```diff
@@ -14,8 +14,129 @@
+
+void DeepDelete(const VkGraphicsPipelineCreateInfo& create_info) {
+  // NOLINTNEXTLINE(cppcoreguidelines-owning-memory)
+  delete[] create_info.pStages;
```
These need to be deep deleted, right?

ilkkasaa replied:
Right! Good catch.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/draw_call_tracker.cc:
```diff
@@ -48,6 +65,37 @@ void DrawCallTracker::HandleDrawCall(uint32_t first_index, uint32_t index_count,
+        graphics_pipeline_data->GetCreateInfo().pStages[stageIndex];
+    if (stage_create_info.stage == VK_SHADER_STAGE_VERTEX_BIT) {
+      vertex_shader =
+          const_cast<VkPipelineShaderStageCreateInfo*>(&stage_create_info);
```
I don't think we want const_cast. I have removed these.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/draw_call_tracker.cc:
```diff
@@ -48,6 +65,37 @@ void DrawCallTracker::HandleDrawCall(uint32_t first_index, uint32_t index_count,
+  }
+  // Both vertex and fragment shaders are required.
+  if (vertex_shader == nullptr || fragment_shader == nullptr) {
+    throw std::runtime_error("Missing vertex or fragment shader.");
```
We probably also don't want to ever use exceptions. I will replace with an assert.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -177,6 +178,99 @@ VKAPI_ATTR void VKAPI_CALL vkCmdDrawIndexed(
+    // Create a ShaderModuleData object to keep track of the shader module's
+    // lifetime.
+    device_data->shader_modules_data.Put(
+        *pShaderModule, std::make_unique<ShaderModuleData>(*pCreateInfo));
```
I have changed this to make_shared, which I guess is what we want?

ilkkasaa replied:
Yeah, right. I forgot to change it when I moved to shared pointers.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -177,6 +178,99 @@ VKAPI_ATTR void VKAPI_CALL vkCmdDrawIndexed(
+  ScopedLock scoped_lock(local_mutex);
+  auto* shader_module_map =
+      device_data->shader_modules_data.Access(&scoped_lock);
+  shader_module_map->erase(shaderModule);
```
I added a Remove method.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/draw_call_tracker.cc:
```diff
@@ -14,17 +14,34 @@
 namespace gf_layers::amber_scoop_layer {
 
+static std::string DisassembleShaderModule(
+    const VkShaderModuleCreateInfo& create_info);
```
I have moved this function up to avoid the forward declaration and used an anonymous namespace instead of using `static`.

paulthomson committed:
A few improvements

paulthomson reviewed (commented):
I pushed a few changes and left a few comments. Mainly, I think there is a missing deep delete.

paulthomson committed:
Remove header

ilkkasaa committed:
Add missing call to DeepDelete()
